src/lib/os-release.c

```diff
@@ -0,0 +1,235 @@
+/*
+ * This file is part of clr-boot-manager.
+ *
+ * Copyright © 2017 Intel Corporation
+ *
+ * clr-boot-manager is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1
+ * of the License, or (at your option) any later version.
+ */
+
+#include "os-release.h"
+#include "log.h"
+#include "nica/files.h"
+
+#include <ctype.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Statically defined fields for correctness of implementation
+ */
+static const char *os_release_fields[] = {
+            [OS_RELEASE_NAME] = "NAME",
+            [OS_RELEASE_VERSION] = "VERSION",
+            [OS_RELEASE_ID] = "ID",
+            [OS_RELEASE_VERSION_ID] = "VERSION_ID",
+            [OS_RELEASE_PRETTY_NAME] = "PRETTY_NAME",
+            [OS_RELEASE_ANSI_COLOR] = "ANSI_COLOR",
+            [OS_RELEASE_HOME_URL] = "HOME_URL",
+            [OS_RELEASE_SUPPORT_URL] = "SUPPORT_URL",
+            [OS_RELEASE_BUG_REPORT_URL] = "BUG_REPORT_URL",
+};
+
+/**
+ * Return a sane fallback key if one isn't provided in the config.
+ */
+static const char *cbm_os_release_fallback_value(CbmOsReleaseKey key)
+{
+        switch (key) {
+        case OS_RELEASE_NAME:
+                return "generic-linux-os";
+        case OS_RELEASE_PRETTY_NAME:
+                return "generic-linux-os";
+        case OS_RELEASE_ID:
+        case OS_RELEASE_VERSION:
+        case OS_RELEASE_VERSION_ID:
+                return "1";
+        default:
+                return "";
+        }
+}
+
+/**
+ * Do the actual hard work of parsing the os-release file.
+ */
+static bool cbm_os_release_parse(CbmOsRelease *self, const char *path)
+{
+        autofree(FILE) *f = NULL;
+        size_t sn;
+        ssize_t r = 0;
+        char *buf = NULL;
+        bool ret = true;
+
+        f = fopen(path, "r");
+        if (!f) {
+                if (errno != ENOENT) {
+                        LOG_ERROR("Unable to open %s: %s", path, strerror(errno));
+                }
+                return false;
+        }
+
+        while ((r = getline(&buf, &sn, f)) > 0) {
+                ssize_t cur = 0;
+                size_t val_len = 0;
+                size_t incr = 0;
+                char *key = NULL;
+                autofree(char) *value = NULL;
```

Comment:
So I thought autofree just works at the function level not the block level. So if you malloc in a loop it will autofree only the last one at function return time.